feat(openchallenges): add temporal information to challenge card

vpchung opened the pull request:
Fixes #2240 

## Changelog
* add logic for calculating time difference in units of years, months, weeks, days, and hours
* display time until start date OR time until end date (depending on status), using the largest non-zero time unit
* due to limited space in the footer: re-arrange the card layout, switching placements of status and platform
* update "Incentives" icon to something more obvious (trophy)
* add fixed heights for status and headline divs; that way, cards stay uniformed in height no matter the contents

## Preview

Footer icons using the same color:

![Screenshot 2023-10-26 at 12 31 22 AM](https://github.com/Sage-Bionetworks/sage-monorepo/assets/9377970/f043f851-941c-4425-9e1e-575f18e086e4)

OR one footer icon uses the blue color from our logo and one icon uses the other:

![Screenshot 2023-10-26 at 4 49 56 PM](https://github.com/Sage-Bionetworks/sage-monorepo/assets/9377970/4c234063-5be0-4494-94d2-0f6273411567)



Alternatively, we can forgo displaying the platform altogether (something Addison mentioned as potentially not helpful information):

![Screenshot 2023-10-26 at 12 44 12 AM](https://github.com/Sage-Bionetworks/sage-monorepo/assets/9377970/cb1ab5c1-4f6d-428b-ba9c-df7b90b9d86f)

vpchung commented:
Latest iteration:

![Screenshot 2023-10-27 at 2 07 44 PM](https://github.com/Sage-Bionetworks/sage-monorepo/assets/9377970/7decd7fc-9c4b-4ec1-b929-a5ff5ba00f70)

tschaffter commented:
@vpchung "Completed" could be reminiscent of errors to users. What about using the default (black) color for completed challenges?

tschaffter commented:
Another suggestion would be to hide time information for challenges that have no longer been "recently" completed (3 mo ago?).

vpchung commented:
### Black color for `completed` status; show all timeSince info
![Screenshot 2023-10-27 at 2 42 25 PM](https://github.com/Sage-Bionetworks/sage-monorepo/assets/9377970/1fa7df74-b882-4e28-b3d2-c7a2d4c5d538)

### Black color for `completed` status; hide timeSince info older than 3 months
![Screenshot 2023-10-27 at 2 50 37 PM](https://github.com/Sage-Bionetworks/sage-monorepo/assets/9377970/2625cfef-7475-488f-a2eb-71e83270d2c7)

### Colorized `completed` status; 
![Screenshot 2023-10-27 at 2 50 58 PM](https://github.com/Sage-Bionetworks/sage-monorepo/assets/9377970/a1c25da3-e100-4895-a53e-b17b7704645b)
hide timeSince info older than 3 months

vpchung commented:
I'm not a big fan of the default color for the `Completed` status, but I can also see your point about red = error.

tschaffter commented:
@vpchung Why are you not a fan of the default color?

vpchung commented:
For me, it lacks consistency - either all statuses should be a different color from the default text color, or they should all be the same.

tschaffter commented:
@vpchung Thanks for the insight. I understand this viewpoint. I gave some thought about this and I think that the new status style capture more the attention that previously. The two reasons are that this information is now at the center of the card and that it's more colorful compared to the solution with the colored dot. With the proposed design, I think that I'm looking first at the status, then at the title. I think that this may disrupt the flow of reading that should be from top to bottom - or at least the title should be read first? Maybe my eyes are attracted to the status first because it's the topic of this PR...

Could we ask Adam to review the current proposal for the card?

@vpchung What do you think about hiding the platform information from the card? Did you mention on Friday that Addison suggested this too? The two reasons that prompt me to suggest this change are: 1) we know that the challenge ecosystem and information are fragmented by benchmarking communities and platforms. Removing this information from the card could be a way for us to "mend" this fragmentation to some extend if that makes sense. 2) enable the user to read more challenge cards by removing some information from the card.

tschaffter commented:
I like the "Slightly bigger bullet size" design.

vpchung commented:
PR is now ready for (re-)review.

tschaffter commented on libs/openchallenges/ui/src/lib/challenge-card/challenge-card.component.ts:
```diff
@@ -63,6 +57,48 @@ export class ChallengeCardComponent implements OnInit {
     }
   }
+
+  calcTimeDiff(date: string) {
```
Declare the return type.

```suggestion
  calcTimeDiff(date: string): string {
```

vpchung replied:
To be addressed in a different PR.

tschaffter replied:
@vpchung Please open a ticket with my comment as Description

tschaffter commented on libs/openchallenges/ui/src/lib/challenge-card/challenge-card.component.ts:
```diff
@@ -63,6 +57,48 @@ export class ChallengeCardComponent implements OnInit {
+
+  calcTimeDiff(date: string) {
+    const pattern = /\d{4}-\d{2}-\d{2}/;
+    if (!pattern.test(date)) {
+      return '';
+    }
```
This function `calcTimeDiff` should validate the input data as you do but validation error should be handled instead of silently returning an empty string.

One approach is to `throw Error(`The input date does not match the schema ${pattern}: ${date}`)`. Then try/catch the function where it is called. The catch block should at least log the error with `console.error()` (note to myself: we should start using a logger within the app).

In the catch block you could then set `timeSince` to `undefined`, which is preferable to an empty string, since an empty since is probably not an expected value for `timeSince`.

Note that a function that may throw an error also has the return type `never`. The return type of the function would then be `string | never`.

Another approach is to change the function `calcTimeDiff` to require a `Date` object as input. The validation would happen where the Date object is constructed. The validation code in `calcTimeDiff` could be removed and the return type would simply be `string`.

The advantage of the second solution is that the function could be reused directly with "native" `Date` objects without the need to do `Date` => `string` => `Date`.

vpchung replied:
To be addressed in a different PR.

vpchung commented:
Thank you, @rrchai , for pointing out the inconsistency of the status colors during the standup.  I have addressed the issue with the following design options:

**Complete as red**
Profile | Card
-- | -- 
![Screenshot 2023-11-01 at 12 22 11 PM](https://github.com/Sage-Bionetworks/sage-monorepo/assets/9377970/cc820518-2df3-46f6-98d6-706bc03371bd) | ![Screenshot 2023-11-01 at 12 22 20 PM](https://github.com/Sage-Bionetworks/sage-monorepo/assets/9377970/9310eb42-9e5b-4948-9216-1cbf7050dbc7)

**Complete as blue**
Profile | Card
-- | -- 
![Screenshot 2023-11-01 at 12 24 17 PM](https://github.com/Sage-Bionetworks/sage-monorepo/assets/9377970/7455e631-a71f-44b9-9dde-f6aea4668246) | ![Screenshot 2023-11-01 at 12 24 28 PM](https://github.com/Sage-Bionetworks/sage-monorepo/assets/9377970/883c7614-d919-4eaa-9edf-45210f1eedf2)

Note that above, I changed the status color on the profile pages from black to red/blue.  The original design is black text, i.e.

![Screenshot 2023-11-01 at 12 35 54 PM](https://github.com/Sage-Bionetworks/sage-monorepo/assets/9377970/0ec3eaf5-4075-4b36-8ecd-fd4106844ad5)

![Screenshot 2023-11-01 at 12 35 24 PM](https://github.com/Sage-Bionetworks/sage-monorepo/assets/9377970/71b17761-bebf-42a1-bbdd-93af27d65f38)

vpchung commented:
Moving forward with blue for "Completed" status. As a safe measure, I also simulated the updated status colors with [Coblis](https://www.color-blindness.com/coblis-color-blindness-simulator/) to verify that this change is still accessible to users with color-blindness (they passed, according to the simulator).

rrchai commented:
LGTM with one small suggestion. Since the status has a circle in front of the text, align-center might not be perfectly center the status box with "ends/starts ..." box. Adding some margin-left on status box to move a bit toward left can make alignment better in my oppinion. 

### Before
<img width="572" alt="Screen Shot 2023-11-01 at 6 49 59 PM" src="https://github.com/Sage-Bionetworks/sage-monorepo/assets/73901500/cd7dd1bc-dff7-4bb6-b6b7-6cd88c8df203">

### After moving status box to left by 8px 
<img width="577" alt="Screen Shot 2023-11-01 at 6 49 43 PM" src="https://github.com/Sage-Bionetworks/sage-monorepo/assets/73901500/1bb26608-6ce0-415b-bf5e-daa8d7164b6b">
